<!doctype html>
<html lang="en">
	<head>
		<meta charset="UTF-8" />
		<title>Call stack, but make it async!</title>
		<link type="image/png" href="../images/favicon.png" rel="shortcut icon" />
		<link type="image/png" href="../images/favicon.png" rel="apple-touch-icon" />
		<link type="text/css" href="../main.css" rel="stylesheet" />
		<meta name="description" content="A deep dive into async stack traces in javascript by Neal Beeken" />
		<meta property="og:url" content="https://nbbeeken.github.io/" />
		<meta property="og:type" content="website" />
		<meta property="og:title" content="Call Stack, But Make It Async!" />
		<meta property="og:description" content="A deep dive into async stack traces in javascript by Neal Beeken" />
		<meta property="og:image" content="../images/js_throw.jpg" />
		<meta name="twitter:card" content="summary_large_image" />
		<meta property="twitter:domain" content="nbbeeken.github.io" />
		<meta property="twitter:url" content="https://nbbeeken.github.io/" />
		<meta name="twitter:title" content="Call Stack, But Make It Async!" />
		<meta name="twitter:description" content="A deep dive into async stack traces in javascript by Neal Beeken" />
		<meta name="twitter:image" content="../images/js_throw.jpg" />
	</head>

	<body>
		<div>
			<input type="checkbox" id="theme-switch" />
			<label for="theme-switch"></label>
		</div>
		<h1>Call Stack, But Make It Async!</h1>
		<blockquote>
			<p>
				Written by Neal Beeken (<a href="https://nbbeeken.github.io/">website</a>,
				<a href="https://github.com/nbbeeken">GitHub</a>)
			</p>
		</blockquote>
		<h1 id="intro">Intro</h1>
		<p>
			In a recent release of the MongoDB Node.js driver (<a
				href="https://github.com/mongodb/node-mongodb-native/releases/tag/v6.5.0"
				>v6.5.0</a
			>) the team completed the effort of getting all our asynchronous operations to report an accurate asynchronous
			stack trace to assist in pinpointing error origination. Here, I&#39;ll walk you through what this feature of
			JavaScript is and how to obtain it at the low price of zero-cost.
		</p>
		<h1 id="calls-and-how-to-stack-them-">Calls and how to stack them üìö</h1>
		<p>
			First, what is a <a href="https://developer.mozilla.org/en-US/docs/Glossary/Call_stack">call stack</a>? A call
			stack is a hidden data structure that stores information about the active subroutines of a program; active
			subroutines being functions that have been called but have yet to complete execution and return control to the
			caller. The main function of the call stack is to keep track of the point to which each active subroutine should
			return control when it finishes executing.
		</p>
		<p>
			Let&#39;s go through an example, take a program that parses a string from its arguments that is an equation like
			&quot;2+2&quot; and computes the result:
		</p>
		<pre>
      main()
        -> parseString()
          -> splitString()
            -> stringLength()
          -> stringToNumber()
        -> add()
        -> printResult()
      -> return;
    </pre>
		<p>
			Most of us are familiar with the above procedural paradigm (whether from JavaScript, C, Java, or Python) where
			each step in the program is synchronous, so our call stack is a clear ordering of dependent procedures. For
			example, if <code>stringLength</code> fails, the call stack would contain <code>stringLength</code>,
			<code>splitString</code>, <code>parseString</code>, and <code>main</code> as active procedures that have yet to
			return to their callers. The error system of our runtime uses this stack trace to generate a helpful error trace:
		</p>
		<pre>
      file://addNumbers.mjs:35
      throw new Error('cannot get string length')
            ^
      Error: cannot get string length
          at stringLength (file://addNumbers.mjs:35:11)
          at splitString (file://addNumbers.mjs:17:17)
          at parseString (file://addNumbers.mjs:11:19)
          at main (file://addNumbers.mjs:4:5)
    </pre>
		<h2 id="the-async-wrench-">The Async wrench üîß</h2>
		<p>
			Everything changes when we shift to an asynchronous programming model, as the introduction of asynchronous work
			means we no longer have strictly dependent procedures. Essentially, async programming is about setting up tasks
			and adding handling that will be invoked some time later when the task is complete.
		</p>
		<p>Let&#39;s add I/O (a read from standard in) into our program to see how this changes our call stack:</p>
		<pre>
      main()
      -> readStdin(handleUserInput)
      // When the user finishes typing
      handleUserInput()
      -> parseString()
        -> splitString()
          -> stringLength()
    </pre>
		<p>
			Now, main&#39;s only job is to ask the runtime to read from stdin and invoke a function of our choice when it is
			done doing so. This means main is no longer an active procedure; it returns leaving it up to the runtime to keep
			the process running until it has input from stdin to hand back to our function <code>handleUserInput</code>.
		</p>
		<p>Here&#39;s what the stack trace looks like:</p>
		<pre>
      file://addNumbers.mjs:42
      throw new Error('cannot get string length')
            ^
      Error: cannot get string length
          at stringLength (file://addNumbers.mjs:42:11)
          at splitString (file://addNumbers.mjs:24:17)
          at parseString (file://addNumbers.mjs:18:19)
          at ReadStream.handleUserInput (file://addNumbers.mjs:11:5)
          at ReadStream.emit (node:events:511:28)
          at addChunk (node:internal/streams/readable:332:12)
          at readableAddChunk (node:internal/streams/readable:305:9)
          at Readable.push (node:internal/streams/readable:242:10)
          at TTY.onStreamRead (node:internal/stream_base_commons:190:23)
    </pre>
		<p>No sign of <code>main</code>, only <code>handleUserInput</code>.</p>
		<p>
			This is a common hazard of asynchronous programming: you are always replacing the record of your active procedures
			as they are all performing task setup that completes and the callbacks they created are later invoked by the
			runtime.
		</p>
		<h1 id="javascript-">JavaScript üíö</h1>
		<p>
			Asynchronous programming has always been at the heart of JS and is one of the central selling points of using
			Node.js.
		</p>
		<p>
			In 2015, the first
			<a href="https://nodejs.org/en/blog/release/v4.2.0">Long Term Support version of Node.js was released</a>, and
			with it came a stable standard library that popularized a common pattern for handling asynchronous tasks. All
			asynchronous tasks would accept a callback as their last argument, with the callback taking at least two
			arguments: an error, and the task&#39;s result. The pattern was that if the first argument was
			<a href="https://developer.mozilla.org/en-US/docs/Glossary/Truthy">truthy</a> (an error object) the task failed,
			and if it was not then the second argument would contain the result.
		</p>
		<p>Here&#39;s a simplified example of a function that reads a file:</p>
		<pre>
      readFile('filename.txt', (error, data) => {
        if (error) {
          console.error(error);
          return;
        }
        console.log('file contents', data);
      });
    </pre>
		<p>
			The Node.js callback pattern is ubiquitous and familiar, resulting in many popular libraries such as the
			<a href="https://www.mongodb.com/docs/drivers/node/current/">MongoDB Node.js driver</a> adopting it as well.
		</p>
		<h2 id="no-throw-only-callback-">No throw, only callback üêï</h2>
		<p>
			<img src="../images/throw.jpg" alt="" />
			<em
				>credit: <a href="https://cupcakelogic.tumblr.com/post/124392369931/she-is-still-learning">cupcakelogic</a></em
			>
		</p>
		<p>
			A challenge associated with the callback pattern is the requirement that the implementer keep in mind execution
			expectations manually, otherwise they can end up with a confusing order of operations.
		</p>
		<p>
			Typically this is something that should be abstracted to the runtime or language, which can be broken down as
			follows:
		</p>
		<p><strong>Error handling</strong></p>
		<p>
			Properly implementing the callback pattern means errors are passed as variables to a chain of handlers so they
			eventually reach the top-level initiator of the async operation. The syntax and keywords
			<code>throw</code>/<code>try</code>/<code>catch</code> can no longer be used for control flow.
		</p>
		<pre>
      try {
        readFile('filename', (error, data) => {
          if (error) {
            /* ? */
          }
        });
      } catch (error) {
        // So what's the truth?
      }
    </pre>
		<p><strong>Runtime order</strong></p>
		<p>
			Callbacks also demand the developers ensure execution order is consistent. If a file is successfully read and the
			contents are returned in the callback passed to <code>readFile</code>, that callback will always run after the
			code that is on the line following <code>readFile</code>. However, say <code>readFile</code> is passed an invalid
			argument, like a number instead of a string for the path. When it invokes the callback with an invalid argument
			error we would still expect that code to run in the same order as the success case:
		</p>
		<table>
			<thead>
				<tr>
					<td>Invocation of <code>readFile</code></td>
					<td>Questionable <code>readFile</code> Implementation</td>
				</tr>
			</thead>
			<tbody>
				<tr>
					<td>
						<pre>
readFile(0xf113, (error, data) => {
  if (error) {
    console.log('cannot read file', error);
    return;
  }
  console.log('contents:', data);
});
console.log('starting to read file');
          </pre
						>
					</td>
					<td>
						<pre>
function readFile(filename, callback) {
  if (typeof filename !== 'string') {
    callback(new Error('invalid argument'));
    return;
  }
  // open & read file ...
}
          </pre
						>
					</td>
				</tr>
			</tbody>
		</table>

		<p>The code above prints:</p>
		<pre>
      cannot read file Error: invalid argument
      starting to read file
    </pre>
		<p>Whereas when I change <code>readFile</code> to be called with a non-existent path:</p>
		<pre>
      starting to read file
      cannot read file Error: /notAPath.txt Does Not Exist
    </pre>
		<p>
			This is unexpected! The implementer of readFile calls the callback synchronously for an invalid type so readFile
			does not return until that callback completes. It is fairly easy to write callback accepting functions that
			inconsistently order their execution in this way.
		</p>
		<h2 id="promises-">Promises ü§û</h2>
		<p>
			Introducing a more structured approach:
			<a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise">Promises</a>. A
			Promise is an object that handles the resolution or rejection of an async operation, mitigating the above issues
			and allowing for many async operations to be chained together without needing to explicitly pass a finalizer
			callback through to each API that would indicate when all tasks are done.
		</p>

		<table>
			<thead>
				<tr>
					<td>Callback Style</td>
					<td>Promise Style</td>
				</tr>
			</thead>
			<tbody>
				<tr>
					<td>
						<pre>
client.connect((error) => {
 if (error) {
   return done(error);
 }
 client
   .db()
   .collection('test')
   .findOne({}, (error, document) => {
     if (error) {
       return done(error);
     }
     console.log(document);
     return done();
   });
});
    </pre
						>
					</td>
					<td>
						<pre>
client
 .connect()
 .then(() => client.db().collection('test').findOne({}))
 .then(document => console.log(document));
 .catch(error => console.error(error));
    </pre
						>
					</td>
				</tr>
			</tbody>
		</table>

		<p>
			Note how in the promise code there is one error handling case as opposed to the two in the callback case. The
			ability to
			<a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#chaining">chain promises</a>
			allows us to treat many async operations as one, the <code>catch</code> handler would be called if either the
			<code>connect</code> or the <code>find</code> methods were to throw an error. This chaining is convenient, but
			when writing JavaScript today we do even better by using special syntax for handling promises.
		</p>
		<h2 id="enter-async-await-">Enter <code>async</code>/<code>await</code> üîÅ</h2>
		<p>
			Mid-2017 JavaScript engines shipped support for <code>async</code>/<code>await</code> syntax allowing programmers
			to write asynchronous operations in a familiar procedural format. Using <code>async</code>/<code>await</code>
			allows the programmer to encode their logical asynchronous dependencies right into the syntax of the language.
		</p>
		<p>
			Let&#39;s return to our user input example, as we can now &quot;await&quot; the input which keeps
			<code>main</code> as the active procedure that began the task to read from standard in.
		</p>
		<blockquote>
			<p>
				&quot;For <code>await</code> the suspend and resume points coincide and so we not only know where we would
				continue, but by coincidence, we also know where we came from.&quot;
			</p>
			<p>
				source:
				<a
					href="https://docs.google.com/document/d/13Sy_kBIJGP0XT34V1CV3nkWya4TwYx9L3Yv45LdGB6Q/edit#heading=h.e6lcalo0cl47"
					>Zero-cost async stack traces</a
				>
			</p>
		</blockquote>
		<p>When the input is available, <code>readStdin</code> will resolve and we can continue with our parsing.</p>
		<pre>
      async main()
        -> await readStdin()
        -> parseString()
    </pre>
		<p>
			When the JavaScript engine reaches the &quot;<code>await</code>&quot;, <code>main</code> is suspended. The engine
			is free to handle other tasks while the read is waiting for our user to type. We can now encode into the syntax of
			the function that it will suspend until some other task completes, and when it continues it maintains the context
			of everything that was in scope when it started.
		</p>
		<pre>
      file://addNumbers.mjs:43
      throw new Error('cannot get string length')
            ^
      Error: cannot get string length
          at stringLength (file://addNumbers.mjs:43:11)
          at splitString (file://addNumbers.mjs:25:17)
          at parseString (file://addNumbers.mjs:19:19)
          at main (file://addNumbers.mjs:9:5)
          at processTicksAndRejections (node:internal/process/task_queues:95:5)
          at async file://addNumbers.mjs:62:1
    </pre>
		<blockquote>
			<p>
				&quot;The fundamental difference between <code>await</code> and manually constructed promises is that
				<code>await X()</code> <strong>suspends</strong> execution of the current function, while
				<code>promise.then(X)</code> will <strong>continue</strong> execution of the current function after adding the
				<code>X</code> call to the callback chain. In the context of stack traces, this difference is pretty
				significant.&quot;
			</p>
			<p>
				source:
				<a href="https://mathiasbynens.be/notes/async-stack-traces">Why await beats Promise#then() ¬∑ Mathias Bynens</a>
			</p>
		</blockquote>
		<h1 id="sample-stack-traces">Sample Stack Traces</h1>
		<pre>
      try {
        await client.connect();
        const document = await client.db().collection('test').findOne({});
        console.log(document);
      } catch (error) {
        console.error(error);
      }
    </pre>
		<p>
			Prior to completing the <code>async</code>/<code>await</code> conversion down to the internal network layer of the
			driver, our error stack would begin at the point of converting a server&#39;s error message into a JavaScript,
			such as:
		</p>
		<pre>
      MongoServerError: Failing command via 'failCommand' failpoint
        at Connection.onMessage (./mongodb/lib/cmap/connection.js:231:30)
        at MessageStream.(anonymous) (./mongodb/lib/cmap/connection.js:61:60)
        at MessageStream.emit (node:events:520:28)
        at processIncomingData (./mongodb/lib/cmap/message_stream.js:125:16)
        at MessageStream._write (./mongodb/lib/cmap/message_stream.js:33:9)
        at writeOrBuffer (node:internal/streams/writable:564:12)
        at _write (node:internal/streams/writable:493:10)
        at Writable.write (node:internal/streams/writable:502:10)
        at Socket.ondata (node:internal/streams/readable:1007:22)
        at Socket.emit (node:events:520:28)
                        ^-- Sadness, that's not my code...
    </pre>
		<p>
			Now, post v6.5.0, the stack trace points directly back to the origination of an operation (we see you
			<code>main.js</code>!):
		</p>
		<pre>
      MongoServerError: Failing command via 'failCommand' failpoint
        at Connection.sendCommand (./mongodb/lib/cmap/connection.js:290:27)
        at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
        at async Connection.command (./mongodb/lib/cmap/connection.js:313:26)
        at async Server.command (./mongodb/lib/sdam/server.js:167:29)
        at async FindOperation.execute (./mongodb/lib/operations/find.js:34:16)
        at async tryOperation (./mongodb/lib/operations/execute_operation.js:192:20)
        at async executeOperation (./mongodb/lib/operations/execute_operation.js:69:16)
        at async FindCursor._initialize (./mongodb/lib/cursor/find_cursor.js:51:26)
        at async FindCursor.cursorInit (./mongodb/lib/cursor/abstract_cursor.js:471:27)
        at async FindCursor.fetchBatch (./mongodb/lib/cursor/abstract_cursor.js:503:13)
        at async FindCursor.next (./mongodb/lib/cursor/abstract_cursor.js:228:13)
        at async Collection.findOne (./mongodb/lib/collection.js:274:21)
        at async main (./mongodb/main.js:19:3)
                       ^-- Yay, that's my code!
    </pre>
		<p><strong>Additional Resources</strong></p>
		<ul>
			<li>
				<a href="https://docs.google.com/document/d/13Sy_kBIJGP0XT34V1CV3nkWya4TwYx9L3Yv45LdGB6Q/edit"
					>Zero-cost async stack traces</a
				>
			</li>
			<li><a href="https://github.com/tc39/proposal-error-stacks">tc39/proposal-error-stacks</a></li>
			<li><a href="https://v8.dev/docs/stack-trace-api">Stack trace API</a></li>
			<li>
				<a href="https://v8.dev/blog/modern-javascript#proper-tail-calls"
					>ES2015, ES2016, and beyond ¬∑ Tail Calls ¬∑ V8</a
				>
			</li>
			<li><a href="https://v8.dev/blog/fast-async">Faster async functions and promises ¬∑ V8</a></li>
			<li>
				<a href="https://mathiasbynens.be/notes/async-stack-traces"
					>Asynchronous stack traces: why await beats Promise#then() ¬∑ Mathias Bynens</a
				>
			</li>
		</ul>
		<script type="module" src="../theme.ts"></script>
	</body>
</html>
